Gradle 8.8 crashed on MacOS 11.x (8.7 still works as expected)

### Current Behavior

Run `./gradlew :sample:app:assembleDebug --stacktrace`
```
GRADLE_OPTS: -Dorg.gradle.jvmargs=-Xmx4g -Dorg.gradle.daemon=false -Dkotlin.incremental=false -XX:+UseParallelGC -XX:+HeapDumpOnOutOfMemoryError -Dfile.encoding=UTF-8
    JAVA_HOME: /Users/runner/hostedtoolcache/Java_Zulu_jdk/17.0.11-9/x[6](https://github.com/hoc081098/kmp-viewmodel/actions/runs/9393781511/job/25870394801#step:7:6)4
    JAVA_HOME_17_X64: /Users/runner/hostedtoolcache/Java_Zulu_jdk/1[7](https://github.com/hoc081098/kmp-viewmodel/actions/runs/9393781511/job/25870394801#step:7:7).0.11-9/x64
    GRADLE_ACTION_ID: gradle/actions/setup-gradle
    GRADLE_BUILD_ACTION_SETUP_COMPLETED: true
    GRADLE_BUILD_ACTION_CACHE_RESTORED: true
    DEVELOCITY_INJECTION_INIT_SCRIPT_NAME: gradle-actions.inject-develocity.init.gradle
    DEVELOCITY_AUTO_INJECTION_CUSTOM_VALUE: gradle-actions
    GITHUB_DEPENDENCY_GRAPH_ENABLED: false


To honour the JVM settings for this build a single-use Daemon process will be forked. For more on this, please refer to https://docs.gradle.org/8.8/userguide/gradle_daemon.html#sec:disabling_the_daemon in the Gradle documentation.
Daemon will be stopped at the end of the build 
Configuration on demand is an incubating feature.
The message received from the daemon indicates that the daemon has disappeared.
Build request sent: Build{id=abe64d9a-d9a5-40b4-a5b8-45620bbdd052, currentDir=/Users/runner/work/kmp-viewmodel/kmp-viewmodel}
Attempting to read last messages from the daemon log...
Daemon pid: 2939
  log file: /Users/runner/.gradle/daemon/8.8/daemon-2939.out.log
----- Last  20 lines from daemon log file - daemon-2939.out.log -----
2024-06-06T01:44:47.402+0000 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Lock acquired on daemon addresses registry.
2024-06-06T01:44:47.404+0000 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Releasing lock on daemon addresses registry.
2024-06-06T01:44:47.405+0000 [DEBUG] [org.gradle.launcher.daemon.server.DaemonStateCoordinator] resetting idle timer
2024-06-06T01:44:47.406+0000 [DEBUG] [org.gradle.launcher.daemon.server.DaemonStateCoordinator] daemon is running. Sleeping until state changes.
2024-06-06T01:44:47.406+0000 [INFO] [org.gradle.launcher.daemon.server.exec.StartBuildOrRespondWithBusy] Daemon is about to start building Build{id=abe64d9a-d9a5-40b4-a5b8-45620bbdd052, currentDir=/Users/runner/work/kmp-viewmodel/kmp-viewmodel}. Dispatching build started information...
2024-06-06T01:44:47.407+0000 [DEBUG] [org.gradle.launcher.daemon.server.SynchronizedDispatchConnection] thread 21: dispatching org.gradle.launcher.daemon.protocol.BuildStarted@680450b8
2024-06-06T01:44:47.410+0000 [DEBUG] [org.gradle.launcher.daemon.server.exec.EstablishBuildEnvironment] Configuring env variables: [PATH, GRADLE_BUILD_ACTION_SETUP_COMPLETED, GITHUB_REPOSITORY_OWNER, GITHUB_REPOSITORY_ID, CHROMEWEBDRIVER, GITHUB_RUN_NUMBER, ANDROID_HOME, NUNIT3_PATH, GITHUB_WORKSPACE, GITHUB_RETENTION_DAYS, GITHUB_API_URL, GITHUB_RUN_ATTEMPT, JAVA_MAIN_CLASS_2874, PWD, GITHUB_JOB, GITHUB_ACTIONS, STATS_V3PS, DOTNET_MULTILEVEL_LOOKUP, NVM_CD_FLAGS, STATS_RDCL, RUNNER_ENVIRONMENT, DEVELOCITY_AUTO_INJECTION_CUSTOM_VALUE, ANDROID_NDK, GITHUB_ENV, STATS_VMD, GITHUB_EVENT_PATH, XPC_FLAGS, STATS_D, LC_ALL, GITHUB_DEPENDENCY_GRAPH_ENABLED, GITHUB_ACTOR_ID, GITHUB_REF_TYPE, LC_CTYPE, SHLVL, RUNNER_TOOL_CACHE, GITHUB_SERVER_URL, JAVA_HOME, GITHUB_REF, RUNNER_NAME, BOOTSTRAP_HASKELL_NONINTERACTIVE, STATS_TRP, XCODE_13_DEVELOPER_DIR, NUNIT_BASE_PATH, XCODE_12_DEVELOPER_DIR, GITHUB_OUTPUT, JAVA_HOME_17_X64, GITHUB_REPOSITORY, GITHUB_STEP_SUMMARY, GITHUB_ACTION_REF, GOROOT_1_21_X64, RCT_NO_LAUNCH_PACKAGER, DEVELOCITY_INJECTION_INIT_SCRIPT_NAME, CI, RUNNER_ARCH, STATS_UE, GITHUB_SHA, ImageOS, ImageVersion, XCODE_11_DEVELOPER_DIR, STATS_VMFE, RUNNER_PERFLOG, PERFLOG_LOCATION_SETTING, RUNNER_TRACKING_ID, RUNNER_WORKSPACE, GRADLE_OPTS, GITHUB_REF_PROTECTED, EDGEWEBDRIVER, GITHUB_WORKFLOW, HOMEBREW_NO_INSTALL_CLEANUP, CONDA, GRADLE_BUILD_ACTION_CACHE_RESTORED, PIPX_HOME, GITHUB_BASE_REF, GITHUB_WORKFLOW_SHA, GITHUB_PATH, ACTIONS_RUNNER_ACTION_ARCHIVE_CACHE, LOGNAME, GITHUB_EVENT_NAME, GITHUB_TRIGGERING_ACTOR, SHELL, AGENT_TOOLSDIRECTORY, HOMEBREW_CASK_OPTS, ANDROID_NDK_HOME, ANDROID_NDK_ROOT, VCPKG_INSTALLATION_ROOT, GITHUB_GRAPHQL_URL, GRADLE_ACTION_ID, JAVA_HOME_21_X64, TMPDIR, RUNNER_OS, DOTNET_ROOT, BOOTSTRAP_HASKELL_INSTALL_NO_STACK_HOOK, GITHUB_WORKFLOW_REF, GITHUB_ACTION, POWERSHELL_DISTRIBUTION_CHANNEL, __CF_USER_TEXT_ENCODING, GITHUB_ACTOR, GOROOT_1_19_X64, PIPX_BIN_DIR, VM_ASSETS, GECKOWEBDRIVER, JAVA_HOME_8_X64, LANG, GITHUB_REPOSITORY_OWNER_ID, STATS_D_D, STATS_EXTP, XPC_SERVICE_NAME, GITHUB_STATE, STATS_EXT, GITHUB_REF_NAME, NVM_DIR, ANDROID_NDK_LATEST_HOME, HOMEBREW_NO_AUTO_UPDATE, GITHUB_ACTION_REPOSITORY, GITHUB_RUN_ID, GOROOT_1_20_X64, USER, ANDROID_SDK_ROOT, SSH_AUTH_SOCK, RUNNER_TEMP, GITHUB_HEAD_REF, JAVA_HOME_11_X64, HOME]
2024-06-06T01:44:47.4[16](https://github.com/hoc081098/kmp-viewmodel/actions/runs/9393781511/job/25870394801#step:7:17)+0000 [DEBUG] [org.gradle.launcher.daemon.server.SynchronizedDispatchConnection] thread 19: received class org.gradle.launcher.daemon.protocol.CloseInput
2024-06-06T01:44:47.4[17](https://github.com/hoc081098/kmp-viewmodel/actions/runs/9393781511/job/25870394801#step:7:18)+0000 [DEBUG] [org.gradle.launcher.daemon.server.DefaultDaemonConnection] thread 19: Received IO message from client: org.gradle.launcher.daemon.protocol.CloseInput@437cdb27
2024-06-06T01:44:47.424+0000 [DEBUG] [org.gradle.launcher.daemon.server.exec.LogToClient] About to start relaying all logs to the client via the connection.
[20](https://github.com/hoc081098/kmp-viewmodel/actions/runs/9393781511/job/25870394801#step:7:21)24-06-06T01:44:47.425+0000 [INFO] [org.gradle.launcher.daemon.server.exec.LogToClient] The client will now receive all logging from the daemon (pid: 2939). The daemon log file: /Users/runner/.gradle/daemon/8.8/daemon-2939.out.log
2024-06-06T01:44:47.427+0000 [DEBUG] [org.gradle.launcher.daemon.server.exec.RequestStopIfSingleUsedDaemon] Requesting daemon stop after processing Build{id=abe64d9a-d9a5-40b4-a5b8-45620bbdd052, currentDir=/Users/runner/work/kmp-viewmodel/kmp-viewmodel}
2024-06-06T01:44:47.428+0000 [DEBUG] [org.gradle.launcher.daemon.server.clientinput.ClientInputForwarder] Closing this process' stdin at end of input.
2024-06-06T01:44:47.428+0000 [LIFECYCLE] [org.gradle.launcher.daemon.server.DaemonStateCoordinator] Daemon will be stopped at the end of the build 
2024-06-06T01:44:47.428+0000 [DEBUG] [org.gradle.launcher.daemon.server.clientinput.ClientInputForwarder] This process will no longer process any forwarded input.
2024-06-06T01:44:47.429+0000 [DEBUG] [org.gradle.launcher.daemon.server.DaemonStateCoordinator] Stop as soon as idle requested. The daemon is busy: true
2024-06-06T01:44:47.431+0000 [DEBUG] [org.gradle.launcher.daemon.server.DaemonStateCoordinator] daemon stop has been requested. Sleeping until state changes.
2024-06-06T01:44:47.431+0000 [DEBUG] [org.gradle.launcher.daemon.server.exec.ExecuteBuild] The daemon has started executing the build.
2024-06-06T01:44:47.432+0000 [DEBUG] [org.gradle.launcher.daemon.server.exec.ExecuteBuild] Executing build with daemon context: DefaultDaemonContext[uid=23327f6e-7775-4474-ac82-769b506c9f8c,javaHome=/Users/runner/hostedtoolcache/Java_Zulu_jdk/17.0.11-9/x64,javaVersion=17,daemonRegistryDir=/Users/runner/.gradle/daemon,pid=2939,idleTimeout=120000,priority=NORMAL,applyInstrumentationAgent=true,nativeServicesMode=ENABLED,daemonOpts=--add-opens=java.base/java.util=ALL-UNNAMED,--add-opens=java.base/java.lang=ALL-UNNAMED,--add-opens=java.base/java.lang.invoke=ALL-UNNAMED,--add-opens=java.prefs/java.util.prefs=ALL-UNNAMED,--add-opens=java.base/java.nio.charset=ALL-UNNAMED,--add-opens=java.base/java.net=ALL-UNNAMED,--add-opens=java.base/java.util.concurrent.atomic=ALL-UNNAMED,-Xmx4g,-Dfile.encoding=UTF-8,-Duser.country=US,-Duser.language=en,-Duser.variant]
Configuration on demand is an incubating feature.
----- End of the daemon log -----

JVM crash log found: file:///Users/runner/work/kmp-viewmodel/kmp-viewmodel/hs_err_pid2939.log

FAILURE: Build failed with an exception.

* What went wrong:
Gradle build daemon disappeared unexpectedly (it may have been killed or may have crashed)

* Try:
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.

* Exception is:
org.gradle.launcher.daemon.client.DaemonDisappearedException: Gradle build daemon disappeared unexpectedly (it may have been killed or may have crashed)
	at org.gradle.launcher.daemon.client.DaemonClient.handleDaemonDisappearance(DaemonClient.java:284)
	at org.gradle.launcher.daemon.client.DaemonClient.monitorBuild(DaemonClient.java:257)
	at org.gradle.launcher.daemon.client.DaemonClient.executeBuild(DaemonClient.java:[21](https://github.com/hoc081098/kmp-viewmodel/actions/runs/9393781511/job/25870394801#step:7:22)2)
	at org.gradle.launcher.daemon.client.SingleUseDaemonClient.execute(SingleUseDaemonClient.java:65)
	at org.gradle.launcher.daemon.client.SingleUseDaemonClient.execute(SingleUseDaemonClient.java:38)
	at org.gradle.launcher.cli.RunBuildAction.run(RunBuildAction.java:57)
	at org.gradle.internal.Actions$RunnableActionAdapter.execute(Actions.java:167)
	at org.gradle.launcher.cli.DefaultCommandLineActionFactory$ParseAndBuildAction.execute(DefaultCommandLineActionFactory.java:261)
	at org.gradle.launcher.cli.DefaultCommandLineActionFactory$ParseAndBuildAction.execute(DefaultCommandLineActionFactory.java:[23](https://github.com/hoc081098/kmp-viewmodel/actions/runs/9393781511/job/25870394801#step:7:24)2)
	at org.gradle.launcher.cli.DebugLoggerWarningAction.execute(DebugLoggerWarningAction.java:74)
	at org.gradle.launcher.cli.DebugLoggerWarningAction.execute(DebugLoggerWarningAction.java:30)
	at org.gradle.launcher.cli.WelcomeMessageAction.execute(WelcomeMessageAction.java:96)
	at org.gradle.launcher.cli.WelcomeMessageAction.execute(WelcomeMessageAction.java:40)
	at org.gradle.launcher.cli.NativeServicesInitializingAction.execute(NativeServicesInitializingAction.java:55)
	at org.gradle.launcher.cli.NativeServicesInitializingAction.execute(NativeServicesInitializingAction.java:29)
	at org.gradle.launcher.cli.ExceptionReportingAction.execute(ExceptionReportingAction.java:41)
	at org.gradle.launcher.cli.ExceptionReportingAction.execute(ExceptionReportingAction.java:[26](https://github.com/hoc081098/kmp-viewmodel/actions/runs/9393781511/job/25870394801#step:7:27))
	at org.gradle.launcher.cli.DefaultCommandLineActionFactory$WithLogging.execute(DefaultCommandLineActionFactory.java:365)
	at org.gradle.launcher.Main.doAction(Main.java:35)
	at org.gradle.launcher.bootstrap.EntryPoint.run(EntryPoint.java:51)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.gradle.launcher.bootstrap.ProcessBootstrap.runNoExit(ProcessBootstrap.java:60)
	at org.gradle.launcher.bootstrap.ProcessBootstrap.run(ProcessBootstrap.java:37)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.gradle.launcher.GradleMain.main(GradleMain.java:34)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.gradle.wrapper.GradleWrapperMain.main(SourceFile:70)

Error: Process completed with exit code 1.
```


### Expected Behavior

Don't crash.
Version 8.7 still works as expected.

### Context (optional)

- Github Action Log: https://github.com/hoc081098/kmp-viewmodel/actions/runs/9393781511/job/25870394801.
- Reproduced on local macOS 11.6.1 machine:
  - MacBook Pro (13-inch, 2019, Four Thunderbolt 3 ports)
  - 2,4 GHz Quad-Core Intel Core i5
  - 16 GB 2133 MHz LPDDR3

- Revert to Gradle 8.7 on MacOS 11, still works.

### Steps to Reproduce

- Clone https://github.com/hoc081098/kmp-viewmodel
- Checkout https://github.com/hoc081098/kmp-viewmodel/pull/286
- Open in Android Studio on MacOS 11.6.1 machine

### Gradle version

8.8

### Build scan URL (optional)

_No response_

### Your Environment (optional)

```
GRADLE_OPTS: -Dorg.gradle.jvmargs=-Xmx4g -Dorg.gradle.daemon=false -Dkotlin.incremental=false -XX:+UseParallelGC -XX:+HeapDumpOnOutOfMemoryError -Dfile.encoding=UTF-8
    JAVA_HOME: /Users/runner/hostedtoolcache/Java_Zulu_jdk/17.0.11-9/x[6](https://github.com/hoc081098/kmp-viewmodel/actions/runs/9393781511/job/25870394801#step:7:6)4
    JAVA_HOME_17_X64: /Users/runner/hostedtoolcache/Java_Zulu_jdk/1[7](https://github.com/hoc081098/kmp-viewmodel/actions/runs/9393781511/job/25870394801#step:7:7).0.11-9/x64
    GRADLE_ACTION_ID: gradle/actions/setup-gradle
    GRADLE_BUILD_ACTION_SETUP_COMPLETED: true
    GRADLE_BUILD_ACTION_CACHE_RESTORED: true
    DEVELOCITY_INJECTION_INIT_SCRIPT_NAME: gradle-actions.inject-develocity.init.gradle
    DEVELOCITY_AUTO_INJECTION_CUSTOM_VALUE: gradle-actions
    GITHUB_DEPENDENCY_GRAPH_ENABLED: false
```

--- a/platforms/core-runtime/native/src/main/java/org/gradle/internal/nativeintegration/services/NativeServices.java
+++ b/platforms/core-runtime/native/src/main/java/org/gradle/internal/nativeintegration/services/NativeServices.java
@@ -58,6 +58,7 @@ import org.gradle.internal.service.Provides;
 import org.gradle.internal.service.ServiceCreationException;
 import org.gradle.internal.service.ServiceRegistration;
 import org.gradle.internal.service.ServiceRegistry;
+import org.gradle.util.internal.VersionNumber;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -515,6 +516,14 @@ public class NativeServices extends DefaultServiceRegistry implements ServiceReg
 
             @Override
             public boolean useFileSystemWatching() {
+                OperatingSystem operatingSystem = OperatingSystem.current();
+                if (operatingSystem.isMacOsX()) {
+                    String version = operatingSystem.getVersion();
+                    if (VersionNumber.parse(version).getMajor() < 12) {
+                        LOGGER.info("Disabling file system watching on macOS {}, as it is only supported for macOS 12+", version);
+                        return false;
+                    }
+                }
                 return isFeatureEnabled(NativeFeatures.FILE_SYSTEM_WATCHING);
             }
         };

--- a/platforms/documentation/docs/src/docs/userguide/optimizing-performance/file_system_watching.adoc
+++ b/platforms/documentation/docs/src/docs/userguide/optimizing-performance/file_system_watching.adoc
@@ -52,7 +52,7 @@ Gradle supports file system watching on the following operating systems:
 ** CentOS Stream 8 or later
 ** Red Hat Enterprise Linux (RHEL) 8 or later
 ** Amazon Linux 2 or later
-* macOS 10.14 (Mojave) or later on Intel and ARM architectures
+* macOS 12 (Monterey) or later on Intel and ARM architectures
 
 == Supported File Systems
 

--- a/platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc
+++ b/platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc
@@ -71,6 +71,11 @@ However, before this release, the configuration cache subsystem did not follow t
 To address this issue, in the current release, all code originally under the `org.gradle.configurationcache*` packages
 (which was never API) has been moved to new internal packages (`org.gradle.internal.*`).
 
+=== File-system watching on macOS 11 (Big Sur) and earlier is disabled
+
+Since Gradle 8.8 file-system watching has only been supported on macOS 12 (Monterey) and later.
+We've now added a check to disable file-system watching on macOS 11 (Big Sur) and earlier automatically.
+
 [[changes_8.8]]
 == Upgrading from 8.7 and earlier
 

